### PR TITLE
Support untitled buffer

### LINKED ISSUE #658: Enable LSP with "No name" buffer
I know Language Server Protocol require file name to manipulate requests/response. But this is not good experience for newbie. How about treat `_NO_NAME_` as file name for it?

### PR DESCRIPTION
**Pull request by mattn**
Fixed #658

### CODE REVIEW

File: plugin/lsp.vim [hunk 1]
```diff
@@ -66,6 +66,7 @@ let g:lsp_tagfunc_source_methods = get(g:, 'lsp_tagfunc_source_methods', ['defin
 let g:lsp_show_message_request_enabled = get(g:, 'lsp_show_message_request_enabled', 1)
 let g:lsp_show_message_log_level = get(g:, 'lsp_show_message_log_level', 'warning')
 let g:lsp_work_done_progress_enabled = get(g:, 'lsp_work_done_progress_enabled', 0)
+let g:lsp_untitled_buffer_enabled = get(g:, 'lsp_untitled_buffer_enabled', 1)
```
Comment (rhysd) [inline plugin/lsp.vim:69]: Can we have documentation for this variable?

File: autoload/lsp/utils.vim [hunk 2]
```diff
@@ -167,7 +167,12 @@ function! lsp#utils#get_buffer_path(...) abort
 endfunction
 
 function! lsp#utils#get_buffer_uri(...) abort
-    return lsp#utils#path_to_uri(expand((a:0 > 0 ? '#' . a:1 : '%') . ':p'))
+    let l:name = expand((a:0 > 0 ? '#' . a:1 : '%') . ':p')
+    if empty(l:name)
+        let l:nr = bufnr(a:0 > 0 ? '#' . a:1 : '%')
```
Comment (rhysd) [inline autoload/lsp/utils.vim:172]: `#` means alternate buffer in parameter of `bufnr()` or `bufname()`. So `#{number}` is not valid format here.

```suggestion
        let l:nr = a:0 > 0 ? a:1 : bufnr('%')
```

would be correct

**Review by rhysd — CHANGES_REQUESTED** [review]

**Comment by rhysd** [comment]
@mattn Just clarification: When opening a new file with file extension, `s:on_text_document_did_open()` will be called twice on `BufNewFile` and `FileType`. Is it ok? For example, `lsp_buffer_enabled` might be notified twice.

**Comment by mattn** [comment]
As far as I looked the part of codes, it seems okay and lsp_buffer_enabled is not triggered twice. But we might need to stop unused servers. (TODO)

**Comment by prabirshrestha** [comment]
what happens if the file is saved with a proper name?

should we unload the unnamed buffer and then load the proper files?

**Comment by mattn** [comment]
When the file is saved, it mean the URI is changed as same as titled file is renamed.

**Review by rhysd — APPROVED** [review]
Code changes look good.

> As far as I looked the part of codes, it seems okay and lsp_buffer_enabled is not triggered twice. But we might need to stop unused servers. (TODO)

Please remember to confirm this before merging this.

**Comment by prabirshrestha** [comment]
LGTM

**Comment by mattn** [comment]
I tried this PR for a few weeks. I had no problems with clangd and some other language servers. Thanks.

**Pull request merged**